jep/0000/README.adoc

```diff
@@ -0,0 +1,232 @@
+     |                           |<--------------------------o
+     |                           o-----------/ok/----------->|
+     |                           |                           |
+     |                     GET /update                       |
+     |<------------------------------------------------------o
+     |    GET /versions for uuid |                           |
+     o-------------------------->|                           |
+     |<-----/200 json payload/---o                           |
+     |                           |                           |
+     |            200 with computed JSON update manifest     |
+     |            indicating what files should               |
+     |            be downloaded                              |
+     o------------------------------------------------------>|
+     |                           |                           |
+----
+
+[[update-manifest]]
+=== Update Manifest
+
+The responses sent to the client must be well-formed JSON documents, referred
+to as "update manifests" which the client must understand. The Update Manifest
+should have a consistent structure but will be dynamically generated _per
+client_ in order to ensure that the client is only downloading what is
+necessary to update that specific client.
+
+.Example Update Manifest
+[source,json]
+----
+{
+    "core" : {
+        "uri" : "https://updates.jenkins.io/some/path/to/a/jenkins.war",
+        "checksum" : {
+            "type" : "sha256",
+            "signature" : "somechecksumforthefile"
+        }
+    },
+    "plugins" : {
+        "updates" : [
+            {
+                "uri" : "https://updates.jenkins.io/some/path/to/a/plugin.hpi",
+                "checksum" : {
+                    "type" : "sha256",
+                    "signature" : "somechecksumforthefile"
+                }
+            },
+            {
+                "uri" : "https://updates.jenkins.io/some/path/to/another/plugin.hpi",
+                "checksum" : {
+                    "type" : "sha256",
+                    "signature" : "somechecksumforthefile"
+                }
+            }
+        ]
+    },
+    "client" : {
+        "uri" : "https://updates.jenkins.io/some/path/to/a/evergreen-client.tar.gz",
+        "checksum" : {
+            "type" : "sha256",
+            "signature" : "somechecksumforthefile"
+        }
+    }
+}
+----
+
+The three primary keys of the update manifest are:
+
+* `core` which indicates that a new jenkins.war is necessary.
+* `plugins` which will include a list of `updates` for plugins. This is an
+  object within the JSON structure rather than a flat array as it is expected
+  that at some point in the future we  may require a `removes` list to properly
+  unpublish legacy or out-dated plugins from instances.
+* `client` which incidates a new tarball for upgrading the `evergreen-client`
+  itself.
+
+Additional keys should be ignored by clients not supporting them to allow the
```

Comment:
A warning in logs would be better than just ignoring unknowned  keys.
It helps to bring attention on something unusual

Reply:
I'm not sure I understand your thoughts on a client-side log warning being useful here. I don't anticipate this being an unusual scenario but rather a scenario where the Update Service is providing more data than the current client knows how to handle, with the assumption that once the client is updated, it will know how to handle it.

Reply:
It's more a general concern about software that voluntary ignore information when it doesn't know how to handle it.

All information in a json should have a meaning especially when dynamically generated.
And if they don't, then they should be removed.

````
Example:
 "As a Jenkins Essential enduser, I expect all packages to be signed so that I am confident that my instance isn't corrupted or tampered"

 During an update, if this message pop up "WARNING: Unkowned key 'signature'"
 This brings valuable information.

 Package signature wasn't checked by the client because:
    1. There is a bug in my client version.
    2. My client version doesn't support this feature yet.
    3. My client version doesn't support this feature anymore.

This is totally different than hiding a key because the client doesn't support it.

So maybe 'warning' isn't an appropriate log level but this information should be accessible somewhere.